File: apps/customer/src/utils/passwordPolicy.ts
/** Rules: ≥6 chars, upper, lower, digit, symbol (non-letter/digit, not whitespace). */

export type PasswordRuleChecks = {
  minLength: boolean;
  hasUpper: boolean;
  hasLower: boolean;
  hasDigit: boolean;
  hasSymbol: boolean;
};

const SYMBOL_RE = /[^A-Za-z0-9\s]/;

export function getPasswordRuleChecks(password: string): PasswordRuleChecks {
  return {
    minLength: password.length >= 6,
    hasUpper: /[A-Z]/.test(password),
    hasLower: /[a-z]/.test(password),
    hasDigit: /[0-9]/.test(password),
    hasSymbol: SYMBOL_RE.test(password),
  };
}

export function passwordMeetsPolicy(password: string): boolean {
  const c = getPasswordRuleChecks(password);
  return c.minLength && c.hasUpper && c.hasLower && c.hasDigit && c.hasSymbol;
}

/** First rule still failing, for submit-time messages. */
export function describePasswordPolicyFailure(password: string): string | null {
  if (!password.trim()) return "Password is required.";
  const c = getPasswordRuleChecks(password);
  if (!c.minLength) return "Use at least 6 characters.";
  if (!c.hasUpper) return "Add at least one uppercase letter (A–Z).";
  if (!c.hasLower) return "Add at least one lowercase letter (a–z).";
  if (!c.hasDigit) return "Add at least one number (0–9).";
  if (!c.hasSymbol) return "Add at least one symbol (e.g. @, #, $, %).";
  return null;
}

File: src/QMS.Api/PasswordPolicy.cs
namespace QMS.Api;

internal static class PasswordPolicy
{
    /// <summary>
    /// Customer password: min 6 chars, uppercase, lowercase, digit, and at least one symbol (non-letter/digit, not whitespace).
    /// </summary>
    public static bool IsValid(string password, out string message)
    {
        if (string.IsNullOrWhiteSpace(password))
        {
            message = "Password is required.";
            return false;
        }

        if (password.Length < 6)
        {
            message = "Password must be at least 6 characters.";
            return false;
        }

        if (!password.Any(char.IsUpper))
        {
            message = "Password must include at least one uppercase letter (A–Z).";
            return false;
        }

        if (!password.Any(char.IsLower))
        {
            message = "Password must include at least one lowercase letter (a–z).";
            return false;
        }

        if (!password.Any(char.IsDigit))
        {
            message = "Password must include at least one number (0–9).";
            return false;
        }

        if (!password.Any(c => !char.IsLetterOrDigit(c) && !char.IsWhiteSpace(c)))
        {
            message = "Password must include at least one symbol (e.g. @, #, $, %).";
            return false;
        }

        message = "";
        return true;
    }
}
